File: JYRadarChart.m
//
//  JYRadarChart.m
//  JYRadarChart
//
//  Created by jy on 13-10-31.
//  Copyright (c) 2013年 wcode. All rights reserved.
//

#import "JYRadarChart.h"
#import "JYLegendView.h"

#define PADDING 13
#define LEGEND_PADDING 3
#define ATTRIBUTE_TEXT_SIZE 10
#define COLOR_HUE_STEP 5
#define MAX_NUM_OF_COLOR 17

@interface JYRadarChart ()

@property (nonatomic, assign) NSUInteger numOfV;
@property (nonatomic, strong) JYLegendView *legendView;
@property (nonatomic, strong) UIFont *scaleFont;

@end

@implementation JYRadarChart

- (id)initWithFrame:(CGRect)frame {
	self = [super initWithFrame:frame];
	if (self) {
        [self setDefaultValues];
	}
	return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self setDefaultValues];
    }
    return self;
}

- (void)setDefaultValues {
    self.backgroundColor = [UIColor whiteColor];
    _maxValue = 100.0;
    _centerPoint = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2);
    _r = MIN(self.frame.size.width / 2 - PADDING, self.frame.size.height / 2 - PADDING);
    _steps = 1;
    _drawPoints = NO;
    _showLegend = NO;
    _showStepText = NO;
    _fillArea = NO;
    _minValue = 0;
    _colorOpacity = 1.0;
    _backgroundLineColor = [UIColor darkGrayColor];
    _backgroundFillColor = [UIColor whiteColor];

    _legendView = [[JYLegendView alloc] init];
    _legendView.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleBottomMargin;
    _legendView.backgroundColor = [UIColor clearColor];
    _legendView.colors = [NSMutableArray array];
    _attributes = @[@"you", @"should", @"set", @"these", @"data", @"titles,",
                        @"this", @"is", @"just", @"a", @"placeholder"];

    _scaleFont = [UIFont systemFontOfSize:ATTRIBUTE_TEXT_SIZE];
}

- (void)setShowLegend:(BOOL)showLegend {
	_showLegend = showLegend;
	if (_showLegend) {
		[self addSubview:self.legendView];
	}
	else {
		for (UIView *subView in self.subviews) {
			if ([subView isKindOfClass:[JYLegendView class]]) {
				[subView removeFromSuperview];
			}
		}
	}
}

- (void)setTitles:(NSArray *)titles {
	self.legendView.titles = titles;
}

- (void)setColors:(NSArray *)colors {
    [self.legendView.colors removeAllObjects];
    for (UIColor *color in colors) {
        [self.legendView.colors addObject:[color colorWithAlphaComponent:self.colorOpacity]];
    }
}

- (void)setNeedsDisplay {
	[super setNeedsDisplay];
	[self.legendView sizeToFit];
	[self.legendView setNeedsDisplay];
}

- (void)setDataSeries:(NSArray *)dataSeries {
	_dataSeries = dataSeries;
	_numOfV = [_dataSeries[0] count];
	if (self.legendView.colors.count < _dataSeries.count) {
		for (int i = 0; i < _dataSeries.count; i++) {
			UIColor *color = [UIColor colorWithHue:1.0 * (i * COLOR_HUE_STEP % MAX_NUM_OF_COLOR) / MAX_NUM_OF_COLOR
			                            saturation:1
			                            brightness:1
			                                 alpha:self.colorOpacity];
			self.legendView.colors[i] = color;
		}
	}
}

- (void)layoutSubviews {
	[self.legendView sizeToFit];
	CGRect r = self.legendView.frame;
	r.origin.x = self.frame.size.width - self.legendView.frame.size.width - LEGEND_PADDING;
	r.origin.y = LEGEND_PADDING;
	self.legendView.frame = r;
	[self bringSubviewToFront:self.legendView];
}

- (void)drawRect:(CGRect)rect {
	NSArray *colors = [self.legendView.colors copy];
	CGFloat radPerV = M_PI * 2 / _numOfV;
	CGContextRef context = UIGraphicsGetCurrentContext();
    
	//draw attribute text
	CGFloat height = [self.scaleFont lineHeight];
	CGFloat padding = 2.0;
	for (int i = 0; i < _numOfV; i++) {
		NSString *attributeName = _attributes[i];
		CGPoint pointOnEdge = CGPointMake(_centerPoint.x - _r * sin(i * radPerV), _centerPoint.y - _r * cos(i * radPerV));
        
		CGSize attributeTextSize = JY_TEXT_SIZE(attributeName, self.scaleFont);
		NSInteger width = attributeTextSize.width;
        
		CGFloat xOffset = pointOnEdge.x >= _centerPoint.x ? width / 2.0 + padding : -width / 2.0 - padding;
		CGFloat yOffset = pointOnEdge.y >= _centerPoint.y ? height / 2.0 + padding : -height / 2.0 - padding;
		CGPoint legendCenter = CGPointMake(pointOnEdge.x + xOffset, pointOnEdge.y + yOffset);

        if (__IPHONE_OS_VERSION_MIN_REQUIRED >= 70000) {
            NSMutableParagraphStyle *paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
            [paragraphStyle setLineBreakMode:NSLineBreakByClipping];
            [paragraphStyle setAlignment:NSTextAlignmentCenter];

            NSDictionary *attributes = @{ NSFontAttributeName: self.scaleFont,
                                          NSParagraphStyleAttributeName: paragraphStyle };

            [attributeName drawInRect:CGRectMake(legendCenter.x - width / 2.0,
                                                 legendCenter.y - height / 2.0,
                                                 width,
                                                 height)
                       withAttributes:attributes];
        }
        else {
            [attributeName drawInRect:CGRectMake(legendCenter.x - width / 2.0,
                                                 legendCenter.y - height / 2.0,
                                                 width,
                                                 height)
                             withFont:self.scaleFont
                        lineBreakMode:NSLineBreakByClipping
                            alignment:NSTextAlignmentCenter];
        }
    }

    //draw background fill color
    [_backgroundFillColor setFill];
    CGContextMoveToPoint(context, _centerPoint.x, _centerPoint.y - _r);
    for (int i = 1; i <= _numOfV; ++i) {
        CGContextAddLineToPoint(context, _centerPoint.x - _r * sin(i * radPerV),
                                _centerPoint.y - _r * cos(i * radPerV));
    }
    CGContextFillPath(context);

	//draw steps line
	//static CGFloat dashedPattern[] = {3,3};
	//TODO: make this color a variable
	[[UIColor lightGrayColor] setStroke];
	CGContextSaveGState(context);
	for (int step = 1; step <= _steps; step++) {
		for (int i = 0; i <= _numOfV; ++i) {
			if (i == 0) {
				CGContextMoveToPoint(context, _centerPoint.x, _centerPoint.y - _r * step / _steps);
			}
			else {
				//                CGContextSetLineDash(context, 0, dashedPattern, 2);
				CGContextAddLineToPoint(context, _centerPoint.x - _r * sin(i * radPerV) * step / _steps,
				                        _centerPoint.y - _r * cos(i * radPerV) * step / _steps);
			}
		}
		CGContextStrokePath(context);
	}
	CGContextRestoreGState(context);
    
	//draw lines from center
	//TODO: make this color a variable
	[[UIColor darkGrayColor] setStroke];
	for (int i = 0; i < _numOfV; i++) {
		CGContextMoveToPoint(context, _centerPoint.x, _centerPoint.y);
		CGContextAddLineToPoint(context, _centerPoint.x - _r * sin(i * radPerV),
		                        _centerPoint.y - _r * cos(i * radPerV));
		CGContextStrokePath(context);
	}
	//end of base except axis label
    
    
	CGContextSetLineWidth(context, 2.0);
    
	//draw lines
	for (int serie = 0; serie < [_dataSeries count]; serie++) {
		if (self.fillArea) {
			[colors[serie] setFill];
		}
		else {
			[colors[serie] setStroke];
		}
		for (int i = 0; i < _numOfV; ++i) {
			CGFloat value = [_dataSeries[serie][i] floatValue];
			if (i == 0) {
				CGContextMoveToPoint(context, _centerPoint.x, _centerPoint.y - (value - _minValue) / (_maxValue - _minValue) * _r);
			}
			else {
				CGContextAddLineToPoint(context, _centerPoint.x - (value - _minValue) / (_maxValue - _minValue) * _r * sin(i * radPerV),
				                        _centerPoint.y - (value - _minValue) / (_maxValue - _minValue) * _r * cos(i * radPerV));
			}
		}
		CGFloat value = [_dataSeries[serie][0] floatValue];
		CGContextAddLineToPoint(context, _centerPoint.x, _centerPoint.y - (value - _minValue) / (_maxValue - _minValue) * _r);
        
		if (self.fillArea) {
			CGContextFillPath(context);
		}
		else {
			CGContextStrokePath(context);
		}
        
        
		//draw data points
		if (_drawPoints) {
			for (int i = 0; i < _numOfV; i++) {
				CGFloat value = [_dataSeries[serie][i] floatValue];
				CGFloat xVal = _centerPoint.x - (value - _minValue) / (_maxValue - _minValue) * _r * sin(i * radPerV);
				CGFloat yVal = _centerPoint.y - (value - _minValue) / (_maxValue - _minValue) * _r * cos(i * radPerV);
                
				[colors[serie] setFill];
				CGContextFillEllipseInRect(context, CGRectMake(xVal - 4, yVal - 4, 8, 8));
				[self.backgroundColor setFill];
				CGContextFillEllipseInRect(context, CGRectMake(xVal - 2, yVal - 2, 4, 4));
			}
		}
	}
    
	if (self.showStepText) {
		//draw step label text, alone y axis
		//TODO: make this color a variable
		[[UIColor blackColor] setFill];
		for (int step = 0; step <= _steps; step++) {
			CGFloat value = _minValue + (_maxValue - _minValue) * step / _steps;
			NSString *currentLabel = [NSString stringWithFormat:@"%.0f", value];
			JY_DRAW_TEXT_IN_RECT(currentLabel,
			                     CGRectMake(_centerPoint.x + 3,
			                                _centerPoint.y - _r * step / _steps - 3,
			                                20,
			                                10),
			                     self.scaleFont);
		}
	}
}

@end

File: JYRadarChartSwiftDemo/JYRadarChartSwiftDemo/JYRadarChart.swift
//
//  JYRadarChart.swift
//  JYRadarChartSwiftDemo
//
//  Created by Erick Santos on 30/04/17.
//
//

import UIKit

class JYRadarChart: UIView {

    private let padding: CGFloat = 13.0
    private let legendPadding: CGFloat = 3.0
    private let atributeTextSize: CGFloat = 10.0
    private let colorHueStep: CGFloat = 5.0
    private let maxNumOfColor: CGFloat = 17.0
    
    lazy var r: CGFloat = {
        [unowned self] in
        min(self.frame.size.width / 2 - self.padding, self.frame.size.height / 2 - self.padding)
        }()
    var maxValue: CGFloat = 100.0
    var minValue: CGFloat = 0.0
    var isDrawPoints: Bool = false
    var isFillArea: Bool = false
    private var isShowTextLegend: Bool = false
    var isShowLegend: Bool {
        get {
            return self.isShowTextLegend
        }
        set {
            self.isShowTextLegend = newValue
            if self.isShowTextLegend {
                addSubview(legendView)
            }
            else {
                for subview in subviews {
                    if subview.isKind(of: JYLegendView.self) {
                        subview.removeFromSuperview()
                    }
                }
            }
        }
    }
    var isShowStepText: Bool = false
    var colorOpacity: CGFloat = 1.0
    var backgroundLineColorRadial: UIColor = UIColor.black
    
    private var dataSeries: Array<Array<Double>> = []
    var data: Array<Array<Double>> {
        get {
            return self.dataSeries
        }
        
        set {
            self.dataSeries = newValue
            self.numOfV = self.dataSeries[0].count
            self.legendView.colors = []
            
            if self.legendView.colors.count < self.dataSeries.count {
                for index in 0..<dataSeries.count {
                    let color: UIColor = UIColor(hue: 1.0 * (CGFloat(index) * self.colorHueStep.truncatingRemainder(dividingBy: self.maxNumOfColor)) / self.maxNumOfColor, saturation: 1.0, brightness: 1.0, alpha: self.colorOpacity)
                    
                    self.legendView.colors.append(color)
                }
            }
        }
    }
    var attributes: Array<String> = []
    var steps: Int = 1
    lazy var centerPoint: CGPoint = {
        [unowned self] in
        return CGPoint(x: self.bounds.size.width / 2, y: self.bounds.size.height / 2)
        }()
    var backgroundFillColor: UIColor = UIColor.white
    
    var isClockWise: Bool = true
    
    var isShowTotal: Bool = true
    private var totalTitleString: String? = nil
    var totalTitle: String {
        get {
            if let totalTitleString = self.totalTitleString {
                return totalTitleString
            } else {
                return String(describing: Int(self.maxValue))
            }
        }
        set {
            self.totalTitleString = newValue
        }
    }
    var titles: Array<String> {
        get {
            return self.legendView.titles
        }
        
        set {
            self.legendView.titles = newValue
        }
    }
    
    var colors: Array<UIColor> {
        get {
            return self.legendView.colors
        }
        
        set {
            self.legendView.colors.removeAll()
            for color in newValue {
                self.legendView.colors.append(color.withAlphaComponent(self.colorOpacity));
            }
        }
    }
    
    var stepLinesColor = UIColor.lightGray
    
    private var numOfV: Int = 0
    private let legendView: JYLegendView = {
        let legendView = JYLegendView()
        legendView.autoresizingMask = [.flexibleLeftMargin, .flexibleBottomMargin]
        legendView.backgroundColor = UIColor.clear
        legendView.colors = []
        return legendView
    }()
    
    var scaleFont: UIFont = UIFont.systemFont(ofSize: 10)
    var fontColors: Array<UIColor> = []
    
    override func setNeedsDisplay() {
        super.setNeedsDisplay()
        
        legendView.sizeToFit()
        legendView.setNeedsDisplay()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        legendView.sizeToFit()
        var r: CGRect = legendView.frame
        r.origin.x = frame.size.width - legendView.frame.size.width - legendPadding
        r.origin.y = legendPadding
        legendView.frame = r
        bringSubview(toFront: self.legendView)
    }
    
    override func draw(_ rect: CGRect) {
        let colors: Array<UIColor> = legendView.colors
        var radPerV: CGFloat = CGFloat(Double.pi * 2.0) / CGFloat(numOfV)
        
        if isClockWise {
            radPerV = -(CGFloat(Double.pi * 2.0) / CGFloat(numOfV))
        }
        
        let context: CGContext? = UIGraphicsGetCurrentContext()
        
        let height: CGFloat = scaleFont.lineHeight
        let padding: CGFloat = 2.0
        
        
        for index in 0..<numOfV {
            let attributeName: String = self.attributes[index]
            let pointOnEdge: CGPoint = CGPoint(x: centerPoint.x - r * sin(CGFloat(index) * radPerV),
                                               y: centerPoint.y - r * cos(CGFloat(index) * radPerV))
            
            let attributeTextSize: CGSize = attributeName.size(attributes: [NSFontAttributeName: scaleFont])
            let width: CGFloat = attributeTextSize.width
            
            var xOffset: CGFloat = width / 2.0 + padding
            
            if pointOnEdge.x <= centerPoint.x {
                xOffset = -width / 2.0 - padding
            }
            
            var yOffset: CGFloat = height / 2.0 + padding
            
            if pointOnEdge.y <= centerPoint.y {
                yOffset = -height / 2.0 - padding
            }
            
            let legendCenter: CGPoint = CGPoint(x: pointOnEdge.x + xOffset, y: pointOnEdge.y + yOffset)
            
            let paragraphStyle: NSMutableParagraphStyle = NSParagraphStyle.default.mutableCopy() as! NSMutableParagraphStyle
            
            paragraphStyle.lineBreakMode = .byClipping
            paragraphStyle.alignment = .center
            
            var fontColor: UIColor = UIColor.black
            
            if fontColors.count == numOfV {
                fontColor = fontColors[index]
            }
            
            var attributes: [String: Any] = [ NSFontAttributeName: scaleFont,
                                              NSParagraphStyleAttributeName: paragraphStyle,
                                              NSForegroundColorAttributeName: fontColor ]
            
            attributeName.draw(in: CGRect(x: legendCenter.x - width / 2.0,
                                          y: legendCenter.y - height / 2.0,
                                          width: width,
                                          height: height),
                               withAttributes: attributes)
            
            if isShowTotal {
                attributes = [ NSFontAttributeName: scaleFont,
                               NSParagraphStyleAttributeName: paragraphStyle,
                               NSForegroundColorAttributeName: UIColor.black ]
                
                totalTitle.draw(in: CGRect(x: legendCenter.x - width / 2.0,
                                           y: (legendCenter.y - height / 2.0) + height,
                                           width: width,
                                           height: height),
                                withAttributes: attributes)
            }
            
        }
        
        backgroundFillColor.setFill()
        context?.move(to: CGPoint(x: centerPoint.x, y: centerPoint.y - r))
        
        for index in 1...numOfV {
            context?.addLine(to: CGPoint(x: centerPoint.x - r * sin(CGFloat(index) * radPerV),
                                         y: centerPoint.y - r * cos(CGFloat(index) * radPerV)))
        }
        
        context?.fillPath()
        
        stepLinesColor.setStroke()
        context?.saveGState()
        
        for step in 1...steps {
            for index in 0...numOfV {
                
                if index == 0 {
                    context?.move(to: CGPoint(x: centerPoint.x,
                                              y: centerPoint.y - r * CGFloat(step) / CGFloat(steps)))
                } else {
                    context?.addLine(to: CGPoint(x: centerPoint.x - r * sin(CGFloat(index) * radPerV) * CGFloat(step) / CGFloat(steps),
                                                 y: centerPoint.y - r * cos(CGFloat(index) * radPerV) * CGFloat(step) / CGFloat(steps)))
                }
                
            }
            
            context?.strokePath()
        }
        
        context?.restoreGState()
        
        backgroundLineColorRadial.setStroke()
        
        for index in 0..<numOfV {
            context?.move(to: CGPoint(x: centerPoint.x, y: centerPoint.y))
            context?.addLine(to: CGPoint(x: centerPoint.x - r * sin(CGFloat(index) * radPerV),
                                         y: centerPoint.y - r * cos(CGFloat(index) * radPerV)))
            context?.strokePath()
        }
        
        context?.setLineWidth(2.0)
        
        if numOfV > 0 {
            for serie in 0..<dataSeries.count {
                
                if isFillArea {
                    colors[serie].setFill()
                } else {
                    colors[serie].setStroke()
                }
                
                for index in 0..<numOfV {
                    let value: CGFloat = CGFloat(dataSeries[serie][index])
                    
                    if index == 0 {
                        context?.move(to: CGPoint(x: centerPoint.x,
                                                  y: centerPoint.y - (value - minValue) / (maxValue - minValue) * r))
                    } else {
                        context?.addLine(to: CGPoint(x: centerPoint.x - (value - minValue) / (maxValue - minValue) * r * sin(CGFloat(index) * radPerV),
                                                     y: centerPoint.y - (value - minValue) / (maxValue - minValue) * r * cos(CGFloat(index) * radPerV)))
                    }
                }
                
                let value: CGFloat = CGFloat(dataSeries[serie][0])
                context?.addLine(to: CGPoint(x: centerPoint.x,
                                             y: centerPoint.y - (value - minValue) / (maxValue - minValue) * r))
                
                if isFillArea {
                    context?.fillPath()
                } else {
                    context?.strokePath()
                }
                
                if isDrawPoints {
                    for index in 0..<numOfV {
                        let value: CGFloat = CGFloat(dataSeries[serie][index])
                        let xVal: CGFloat = centerPoint.x - (value - minValue) / (maxValue - minValue) * r * sin(CGFloat(index) * radPerV)
                        let yVal: CGFloat = centerPoint.y - (value - minValue) / (maxValue - minValue) * r * cos(CGFloat(index) * radPerV)
                        
                        colors[serie].setFill()
                        context?.fillEllipse(in: CGRect(x: xVal - 4, y: yVal - 4, width: 8, height: 8))
                        backgroundColor?.setFill()
                        context?.fillEllipse(in: CGRect(x: xVal - 2, y: yVal - 2, width: 4, height: 4))
                    }
                }
                
            }
        }
        
        if isShowStepText {
            
            UIColor.black.setFill()
            for step in 0...steps {
                let value: CGFloat = minValue + (maxValue - minValue) * CGFloat(step) / CGFloat(steps)
                let currentLabel: String = String(describing: Int(value))
                
                currentLabel.draw(in: CGRect(x: centerPoint.x + 3,
                                             y: centerPoint.y - r * CGFloat(step) / CGFloat(steps),
                                             width: 20,
                                             height: 10),
                                  withAttributes: [ NSFontAttributeName: scaleFont ])
            }
            
        }
        
    }

}
